worlds/mm2/__init__.py

```diff
@@ -0,0 +1,311 @@
+                menu.connect(stage, f"To {region}",
+                             lambda state, items=required_items: state.has_all(items, self.player))
+            else:
+                old_stage = self.get_region(prev_stage)
+                old_stage.connect(stage, f"To {region}",
+                                  lambda state, items=required_items: state.has_all(items, self.player))
+            stage.add_locations(locations, MM2Location)
+            for location in stage.get_locations():
+                if location.address is None and location.name != dr_wily:
+                    location.place_locked_item(MM2Item(location.name, ItemClassification.progression,
+                                                       None, self.player))
+            if region in etank_1ups and self.options.consumables in (Consumables.option_1up_etank,
+                                                                     Consumables.option_all):
+                stage.add_locations(etank_1ups[region], MM2Location)
+            if region in energy_pickups and self.options.consumables in (Consumables.option_weapon_health,
+                                                                         Consumables.option_all):
+                stage.add_locations(energy_pickups[region], MM2Location)
+            self.multiworld.regions.append(stage)
+
+    def create_item(self, name: str) -> MM2Item:
+        item = item_table[name]
+        classification = ItemClassification.filler
+        if item.progression:
+            classification = ItemClassification.progression_skip_balancing \
+                if item.skip_balancing else ItemClassification.progression
+        if item.useful:
+            classification |= ItemClassification.useful
+        return MM2Item(name, classification, item.code, self.player)
+
+    def get_filler_item_name(self) -> str:
+        return self.random.choices(list(filler_item_weights.keys()),
+                                              weights=list(filler_item_weights.values()))[0]
+
+    def create_items(self) -> None:
+        itempool = []
+        # grab first robot master
+        robot_master = self.item_id_to_name[0x880101 + self.options.starting_robot_master.value]
+        self.multiworld.push_precollected(self.create_item(robot_master))
+        itempool.extend([self.create_item(name) for name in stage_access_table.keys()
+                         if name != robot_master])
+        itempool.extend([self.create_item(name) for name in robot_master_weapon_table.keys()])
+        itempool.extend([self.create_item(name) for name in item_item_table.keys()])
+        total_checks = 24
+        if self.options.consumables in (Consumables.option_1up_etank,
+                                        Consumables.option_all):
+            total_checks += 20
+        if self.options.consumables in (Consumables.option_weapon_health,
+                                        Consumables.option_all):
+            total_checks += 27
```

Comment:
Idk if you plan to ever update this game, but this seems like you should just use `len(self.multiworld.get_unfilled_locations(self.player))`

Reply:
I think that would actually return one too many locations (since we fill victory in generate_basic), but regardless I don't think there's actually anymore locations we can add. We're stretched thin as is.